Ltac2 preterm quotation and notation scope (#16740)
coq/coq: doc/sphinx/proof-engine/ltac2.rst

Strict vs. non-strict mode
++++++++++++++++++++++++++

> Depending on the context, quotation-producing terms (i.e. ``constr``,
> ``open_constr`` or ``preterm``) are not internalized in the same way.
> There are two possible modes, the *strict* and the *non-strict* mode.

Reviewer: ```suggestion
Terms can be internalized in two ways: *strict* and *non-strict*, depending
on the context.
```

Then explain how  ``constr``, ``open_constr`` and ``preterm`` are handled after the explanation of the modes/ways.

I'd also change "mode" to "way" in this subsection because it's not a settable mode, is it?

Author: I don't see how that would be better.

Reviewer: "not internalized in the same way" is an awkward way of saying that they are internalized differently.  Since you mention there is a difference, it should be described briefly somewhere.  How are users supposed to understand which constructs are strict or non-strict?  How do know what strict and non-strict mean?

Author: todo for some other PR, make an issue if you want to track it

Reviewer: Track it, as in you'll do it later?  If you know the technical details, you should spend the few minutes needed now to briefly document them.  Poorly described functionality probably won't be used.  Everyone needs to take some responsibility for the documentation.  I certainly can't do all of it and I usually need help from others to get details right.

Author: As in I'm not doing it in this PR and I'm not commiting to doing it in the future.

Reviewer 2: I think you are talking about different things.  @jfehrle, "``constr``,
``open_constr`` or ``preterm``" is a definition of what it means for an Ltac2 term to be "quotation-producing".  If you write `true` or `()` or `f x` in Ltac2, these are not quotation-producing.  The Ltac2 terms that are quotation-producing are the terms that are interpreted in Ltac2-scope constr or open_constr.  This PR additionally adds preterm scope, which is also quotation producing.  This section is not about the differences between constr, open_constr, and preterm; this section is about the difference between strict and non-strict mode.  The difference is [already described in the existing documentation](https://coq.inria.fr/refman/proof-engine/ltac2.html#strict-vs-non-strict-mode).  This PR is not changing anything of significance here, just adding "preterm" to the list of quotation-producing internalization scopes, and so this PR is not the proper place to make wording suggestions about explaining strict mode.

The differences between constr, open_constr, and preterm are described above:
```
- ``open_constr``, which parses Coq terms and produces a term potentially with
  holes at runtime (type ``Init.constr`` as well).
- ``preterm``, which parses Coq terms and produces a value which must
  be typechecked with ``Constr.pretype`` (type ``Init.preterm``).
```
Here "parsing a term and producing a term/value" is a roughly synonymous with "internalization"

Reviewer: Ah, good point. Then it would make sense to hyperlink to the "Strict vs. non-strict mode" section.  Either as a `:ref:` or add the two terms to the glossary (`:gdef:`/`:term:`).

Reviewer 2: @jfehrle This change is itself of the first paragraph of the strict vs non-strict section.

Reviewer: OK, my bad, nevermind.

    Parses an open :token:`term`. Like :n:`constr` above, this class
    accepts a list of notation scopes with the same effects.

>   :n:`preterm {? ( {+, @scope_key } ) }`
>     Parses a non-typechecked :token:`term`. Like :n:`constr` above, this class
>     accepts a list of notation scopes with the same effects.

Reviewer: ```suggestion
  :n:`preterm {? ( {+, @scope_key } ) }`
    Parses a :token:`term` without immediately typechecking it. Like :n:`constr` above, this class
    accepts a list of notation scopes with the same effects.
```

Would you add a sentence explaining when the term would eventually be typechecked and why/when `preterm` is useful?  Maybe add a hyperlink to `preterm` (which I suggest above)?

"non-typechecked" seems imprecise; eventually it will be typechecked

Reviewer 2: It is typechecked only when explicitly passed to `Constr.pretype`.  it's useful when you want to add more information before typechecking; for example, when you write `exact foo`, `foo` should not be typechecked immediately; instead `foo : <the type of the goal>` should be typechecked.

Author: Currently I'm not sure preterm is useful as it has too little API.

>"non-typechecked" seems imprecise; eventually it will be typechecked

Not if we don't call the typechecking API on it. Non-typechecked is fine IMO

Reviewer 3: @SkySkimmer you get preterms in notations.

Reviewer: > Non-typechecked is fine IMO

The problem is that AFAIK the doc doesn't say anything about when typechecking is done or skipped entirely or why this is a good thing.  As a non-expert, it seems strange that some things are not typechecked.  It would be a great improvement if you could provide a few sentences that describe when typechecking is done and when not.  Or if you give me the information in a couple bullet points, I'll work on the wording.

Author: Currently it is typechecked by passing to Constr.pretype. In the future there may be other APIs that do it.

Reviewer 2: > it seems strange that some things are not typechecked

Typechecking is costly and slow, and may fail for ill-typed or incomplete terms that can nonetheless hold useful data about how to build well-typed terms.  It's useful to have access to this.
